{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# This is a script for calculating rank product\n",
    "#### Rand product can be used to get the rank of each LR for each i.e. subtype, using multiple samples\n",
    "#### Using MNL COEFFICIENT"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"Calculate rank product from scRNA-seq fold-change data.\"\"\"\n",
    "\"\"\"Implimented from Christie's script\"\"\"\n",
    "\n",
    "import yaml\n",
    "import sys\n",
    "import pandas as pd\n",
    "from scipy.stats.mstats import gmean\n",
    "import os\n",
    "import re"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [],
   "source": [
    "def rank_product(foldchange: pd.DataFrame, n_permutations: int = 100) -> pd.DataFrame:\n",
    "    \"\"\"Calculate rank product from differential expression fold change data.\n",
    "\n",
    "    Args:\n",
    "        foldchange (pd.DataFrame): Fold change values for every detected gene\n",
    "        (columns) per sample (rows). \n",
    "        n_permutations (int, optional): Number of permutations with which to\n",
    "        calculate significance levels. Defaults to 100.\n",
    "\n",
    "    Returns:\n",
    "        pd.DataFrame: Original fold change data, appended with additional\n",
    "        columns for geometric mean of ranks (geo_mean), rank product value\n",
    "        (rank), and permutation-based estimate of percentage of false positives\n",
    "        for the calculated rank product value (pVal). Rows are sorted by\n",
    "        ascending rank product. \n",
    "    \"\"\"\n",
    "    # Calculate the rank product of fold change values by ranking the geometric\n",
    "    # mean of ranked values across samples.\n",
    "    rank = foldchange.rank(axis=0, ascending=False)\n",
    "    rank[\"geo_mean\"] = gmean(rank, axis=1, nan_policy=\"omit\")\n",
    "    rank.sort_values(\"geo_mean\", inplace=True)\n",
    "    rank[\"rank\"] = range(1, len(rank) + 1)\n",
    "    # Calculate the significance for each rank product value by permuting the\n",
    "    # fold change values randomly, calculating the rank product, and comparing\n",
    "    # how likely the given value or better is observed.\n",
    "    shuffle = rank.drop([\"geo_mean\", \"rank\"], axis=1).copy()\n",
    "    c = pd.DataFrame(index=shuffle.index)\n",
    "    for k in range(n_permutations):\n",
    "        for col in shuffle.columns:\n",
    "            shuffle[col] = shuffle[col].sample(frac=1).values\n",
    "        geo_mean = gmean(shuffle, axis=1)\n",
    "        c[k] = geo_mean <= rank[\"geo_mean\"]\n",
    "    erp = c.sum(axis=1) / n_permutations\n",
    "    pfp = erp / rank[\"rank\"]\n",
    "    rank[\"pVal\"] = pfp\n",
    "\n",
    "    return rank\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load MNL coefficient files\n",
    "directory_path = '/Users/victoriagao/local_docs/NEST/stored_variables/MNL_subtype_coefficients/3.5/'\n",
    "file_names = ['new_aggregated_coefficients_pvalues_for_BasalB.csv', 'new_aggregated_coefficients_pvalues_for_ClassicA.csv']\n",
    "# file_path = [os.path.join(directory_path, file_name) for file_name in file_names]\n",
    "\n",
    "data = {}\n",
    "pattern = re.compile(r'for_(.*?)\\.csv')\n",
    "\n",
    "# Loop through each file, extract data, and populate the dictionary\n",
    "for file_name in file_names:\n",
    "    file_path = os.path.join(directory_path, file_name)\n",
    "    df = pd.read_csv(file_path)[['ligand-receptor', 'Coefficient']]\n",
    "    # Convert the DataFrame to a dictionary with 'ligand-receptor' as keys and 'coefficient' as values\n",
    "    data_dict = pd.Series(df.Coefficient.values, index=df['ligand-receptor']).to_dict()\n",
    "    \n",
    "    label = pattern.search(file_name).group(1)\n",
    "    # Add this dictionary to the main data dictionary, using a label for the file as the key\n",
    "    data[label] = data_dict\n",
    "\n",
    "df_final = pd.DataFrame.from_dict(data, orient='index')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>HLA-DRA</th>\n",
       "      <th>GSTP1-IL7R</th>\n",
       "      <th>PSAP-ERBB2</th>\n",
       "      <th>CALR-ITGA5</th>\n",
       "      <th>C3-LPAR2</th>\n",
       "      <th>APP-TNFRSF14</th>\n",
       "      <th>MIF-CD74</th>\n",
       "      <th>THBS1-LRP5</th>\n",
       "      <th>CELSR1-HLA</th>\n",
       "      <th>ANXA1-F2R</th>\n",
       "      <th>...</th>\n",
       "      <th>GDF15-ERBB2</th>\n",
       "      <th>HSP90B1-ERBB2</th>\n",
       "      <th>TNFSF12-TNFRSF12A</th>\n",
       "      <th>C3-FPR1</th>\n",
       "      <th>LGALS3-NPTN</th>\n",
       "      <th>FN1-IGF2R</th>\n",
       "      <th>PKD1-CFTR</th>\n",
       "      <th>GDF15-TGFBR2</th>\n",
       "      <th>APOE-SDC2</th>\n",
       "      <th>ITGB1-ITGA11</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>BasalB</th>\n",
       "      <td>0.308893</td>\n",
       "      <td>0.277297</td>\n",
       "      <td>0.268669</td>\n",
       "      <td>0.246829</td>\n",
       "      <td>0.234998</td>\n",
       "      <td>0.212181</td>\n",
       "      <td>0.207838</td>\n",
       "      <td>0.190971</td>\n",
       "      <td>0.189419</td>\n",
       "      <td>0.180496</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.134090</td>\n",
       "      <td>-0.136220</td>\n",
       "      <td>-0.155665</td>\n",
       "      <td>-0.157654</td>\n",
       "      <td>-0.159110</td>\n",
       "      <td>-0.160355</td>\n",
       "      <td>-0.199030</td>\n",
       "      <td>-0.199778</td>\n",
       "      <td>-0.221654</td>\n",
       "      <td>-0.258492</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ClassicA</th>\n",
       "      <td>-0.036986</td>\n",
       "      <td>-0.070131</td>\n",
       "      <td>-0.020469</td>\n",
       "      <td>0.010304</td>\n",
       "      <td>-0.201364</td>\n",
       "      <td>-0.117449</td>\n",
       "      <td>-0.202332</td>\n",
       "      <td>-0.045784</td>\n",
       "      <td>-0.142147</td>\n",
       "      <td>-0.026344</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.046564</td>\n",
       "      <td>0.067896</td>\n",
       "      <td>0.191911</td>\n",
       "      <td>0.031672</td>\n",
       "      <td>0.003202</td>\n",
       "      <td>-0.003834</td>\n",
       "      <td>0.081375</td>\n",
       "      <td>0.063486</td>\n",
       "      <td>-0.016156</td>\n",
       "      <td>-0.119712</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2 rows Ã— 468 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           HLA-DRA  GSTP1-IL7R  PSAP-ERBB2  CALR-ITGA5  C3-LPAR2  \\\n",
       "BasalB    0.308893    0.277297    0.268669    0.246829  0.234998   \n",
       "ClassicA -0.036986   -0.070131   -0.020469    0.010304 -0.201364   \n",
       "\n",
       "          APP-TNFRSF14  MIF-CD74  THBS1-LRP5  CELSR1-HLA  ANXA1-F2R  ...  \\\n",
       "BasalB        0.212181  0.207838    0.190971    0.189419   0.180496  ...   \n",
       "ClassicA     -0.117449 -0.202332   -0.045784   -0.142147  -0.026344  ...   \n",
       "\n",
       "          GDF15-ERBB2  HSP90B1-ERBB2  TNFSF12-TNFRSF12A   C3-FPR1  \\\n",
       "BasalB      -0.134090      -0.136220          -0.155665 -0.157654   \n",
       "ClassicA    -0.046564       0.067896           0.191911  0.031672   \n",
       "\n",
       "          LGALS3-NPTN  FN1-IGF2R  PKD1-CFTR  GDF15-TGFBR2  APOE-SDC2  \\\n",
       "BasalB      -0.159110  -0.160355  -0.199030     -0.199778  -0.221654   \n",
       "ClassicA     0.003202  -0.003834   0.081375      0.063486  -0.016156   \n",
       "\n",
       "          ITGB1-ITGA11  \n",
       "BasalB       -0.258492  \n",
       "ClassicA     -0.119712  \n",
       "\n",
       "[2 rows x 468 columns]"
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_final"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>BasalB</th>\n",
       "      <th>ClassicA</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>HLA-DRA</th>\n",
       "      <td>0.308893</td>\n",
       "      <td>-0.036986</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSTP1-IL7R</th>\n",
       "      <td>0.277297</td>\n",
       "      <td>-0.070131</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PSAP-ERBB2</th>\n",
       "      <td>0.268669</td>\n",
       "      <td>-0.020469</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CALR-ITGA5</th>\n",
       "      <td>0.246829</td>\n",
       "      <td>0.010304</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C3-LPAR2</th>\n",
       "      <td>0.234998</td>\n",
       "      <td>-0.201364</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FN1-IGF2R</th>\n",
       "      <td>-0.160355</td>\n",
       "      <td>-0.003834</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PKD1-CFTR</th>\n",
       "      <td>-0.199030</td>\n",
       "      <td>0.081375</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GDF15-TGFBR2</th>\n",
       "      <td>-0.199778</td>\n",
       "      <td>0.063486</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>APOE-SDC2</th>\n",
       "      <td>-0.221654</td>\n",
       "      <td>-0.016156</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ITGB1-ITGA11</th>\n",
       "      <td>-0.258492</td>\n",
       "      <td>-0.119712</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>468 rows Ã— 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                BasalB  ClassicA\n",
       "HLA-DRA       0.308893 -0.036986\n",
       "GSTP1-IL7R    0.277297 -0.070131\n",
       "PSAP-ERBB2    0.268669 -0.020469\n",
       "CALR-ITGA5    0.246829  0.010304\n",
       "C3-LPAR2      0.234998 -0.201364\n",
       "...                ...       ...\n",
       "FN1-IGF2R    -0.160355 -0.003834\n",
       "PKD1-CFTR    -0.199030  0.081375\n",
       "GDF15-TGFBR2 -0.199778  0.063486\n",
       "APOE-SDC2    -0.221654 -0.016156\n",
       "ITGB1-ITGA11 -0.258492 -0.119712\n",
       "\n",
       "[468 rows x 2 columns]"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_pivoted = df_final.transpose()\n",
    "df_pivoted"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [],
   "source": [
    "def main():\n",
    "    coefficient_df = df_pivoted\n",
    "    rank_df = rank_product(coefficient_df)\n",
    "    rank_df.to_csv(f\"{directory_path}new_rank_product.csv\")\n",
    "\n",
    "main()\n",
    "\n",
    "# if __name__ == '__main__':\n",
    "#     sys.exit(main())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "interpreter": {
   "hash": "b5a1b890bcc132e984968415f52b57656df096488cf2afb066f7783daeca48fd"
  },
  "kernelspec": {
   "display_name": "Python 3.11.5 64-bit ('base': conda)",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.5"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
